### PR TITLE
Rule for parameter name in outer lambda 

### LINKED ISSUE #1009: Lambda argument name
Create a rule to name outer lambda's argument

## Before
```kotlin 
x.map { 
   it.map { node ->
       node.foo()
   }
}
```

## After
```kotlin 
x.map { node -> 
   node.map { it.foo()  }
}
```

### PR DESCRIPTION
**Pull request by nulls**
## Which rule and warnings did you add?

This pull request closes #1009 
Warning on using implicit parameter name in outer lambda

## Actions checklist
* [x] Implemented Rule, added Warnings
* [x] Added tests on checks
* [x] ~Added tests on fixers~ not applicable
* [x] Updated diktat-analysis.yml
* [x] Updated available-rules.md

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    private fun checkLambda(node: ASTNode) {
+        val copyNode = node.clone() as ASTNode
+
+        val lambdaCount = copyNode.findAllNodesWithCondition { it.elementType == ElementType.LAMBDA_EXPRESSION }.size
+        if (lambdaCount > 1) {
+            copyNode.findAllNodesWithCondition { it.elementType == ElementType.LAMBDA_EXPRESSION }
+                    .forEachIndexed { index, astNode ->
+                        if (index > 0) {
+                            astNode.treeParent.removeChild(astNode)
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt:34]: I thought that this inspection is not intended to make changes in a tree (fixes). Why is that needed?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            checkLambda(node)
+        }
+    }
+
+    private fun checkLambda(node: ASTNode) {
+        val copyNode = node.clone() as ASTNode
+
+        val lambdaCount = copyNode.findAllNodesWithCondition { it.elementType == ElementType.LAMBDA_EXPRESSION }.size
+        if (lambdaCount > 1) {
+            copyNode.findAllNodesWithCondition { it.elementType == ElementType.LAMBDA_EXPRESSION }
```
Comment (kgevorkyan) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt:31]: Yeah, let's extract and rework this logic  - add comments, make sure that we really need some `copyNodes`, which we for some reasons modify

For now it not really clear for me, what we do here
We looking for existence of `it` keyword and looking for some `VALUE_PARAMETER_LIST`, what for?

File: diktat-analysis.yml [hunk 3]
```diff
@@ -429,7 +429,10 @@
 # Checks that using unnecessary, custom label
 - name: CUSTOM_LABEL
   enabled: true
-# Checks that property in constructor doesn't contains comment
+# Checks that outer lambda has explicit parameter name
```
Comment (kgevorkyan) [inline diktat-analysis.yml:432]: I think we also need to update other `diktat-analysis.yml` files, that we have, some of them used for smoke tests, some need for internal usage

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                    .forEachIndexed { index, astNode ->
+                        if (index > 0) {
+                            astNode.treeParent.removeChild(astNode)
+                        }
+                    }
+            val isIt = copyNode.findAllDescendantsWithSpecificType(ElementType.REFERENCE_EXPRESSION).map { re -> re.text }.contains("it")
```
[the remark this answers is not in the document]
Comment (nulls) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt:37]: agree, @kgevorkyan @akuleshov7 do you have an util class for such common methods?

File: diktat-analysis.yml (hunk 3 above)
Comment (akuleshov7) [reply diktat-analysis.yml:432]: Yes, there is a bug after the update to new Kotlin, that caused this unit test (that checked that all diktat-analysis.yml were updated) to be ignored https://github.com/analysis-dev/diktat/issues/1222

I will investigate it one day...

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (akuleshov7) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt:37]: No, we don't have, better to move it to common AstNodeUtils file. 
Awesome thing here, is that Kotlin will help you to create an extension function for this logic.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    private fun checkLambda(node: ASTNode) {
+        val copyNode = node.clone() as ASTNode
+
+        val lambdaCount = copyNode.findAllNodesWithCondition { it.elementType == ElementType.LAMBDA_EXPRESSION }.size
+        if (lambdaCount > 1) {
+            copyNode.findAllNodesWithCondition { it.elementType == ElementType.LAMBDA_EXPRESSION }
+                    .forEachIndexed { index, astNode ->
+                        if (index > 0) {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt:33]: `.drop(1)` can be used instead of index comparison

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt (hunk 4 above)
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt:37]:
```suggestion
            val isIt = copyNode.findAllDescendantsWithSpecificType(ElementType.REFERENCE_EXPRESSION).map { it.text }.contains("it")
```
Usually in Kotlin named arguments aren't used in such short lambdas :)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+                            astNode.treeParent.removeChild(astNode)
+                        }
+                    }
+            val isIt = copyNode.findAllDescendantsWithSpecificType(ElementType.REFERENCE_EXPRESSION).map { re -> re.text }.contains("it")
+            val parameters = node.findChildByType(ElementType.FUNCTION_LITERAL)?.findChildByType(ElementType.VALUE_PARAMETER_LIST)
+            if (parameters == null && isIt) {
+                PARAMETER_NAME_IN_OUTER_LAMBDA.warn(configRules, emitWarn, isFixMode,
+                        "lambda without arguments has inner lambda",
+                        node.startOffset, node)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter5/ParameterNameInOuterLambdaRule.kt:42]: For better usability, offset of the first node with `it` can be used. Now user will be pointed to the very beginning of the top-level lambda.

File: diktat-analysis.yml (hunk 3 above)
Comment (petertrr) [reply diktat-analysis.yml:432]: That issue is already fixed, and once @nulls pulls new changes from master (actually, Github is experiencing degraded performance right now, so the button `Update branch` is missing on this PR 😅), that test should get triggered.

Comment (akuleshov7) [reply diktat-analysis.yml:432]: fixed and merged. Thanks to @Cheshiriks

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt [hunk 7]
```diff
@@ -158,6 +158,7 @@ enum class Warnings(
     RUN_BLOCKING_INSIDE_ASYNC(false, "5.2.4", "avoid using runBlocking in asynchronous code"),
     TOO_MANY_LINES_IN_LAMBDA(false, "5.2.5", "long lambdas should have a parameter name instead of it"),
     CUSTOM_LABEL(false, "5.2.6", "avoid using expression with custom label"),
+    PARAMETER_NAME_IN_OUTER_LAMBDA(false, "5.2.7", "outer lambdas should have a parameter name instead of it"),
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/constants/Warnings.kt:161]:
```suggestion
    PARAMETER_NAME_IN_OUTER_LAMBDA(false, "5.2.7", "outer lambdas should have a parameter name instead of `it`"),
```

**Review by petertrr — COMMENTED** [review]
LGTM, but I've just approved CI workflows, so let's see if everything passes...

**Comment by akuleshov7** [comment]
> @nulls I've fixed some minor issues in your PR (`countCodeLines` was still mutating the AST in `check` mode) and disabled your new inspection for our self style check. It has found 34 issues and it's better we fix them in a separate PR 😅 Also, because you've changed the logic around lambda length, there appeared some false positives of `LambdaLength` rule, e.g. [Code scanning alert](/analysis-dev/diktat/security/code-scanning/85). The inspection is triggered when explicit lambda arguments are named `it`. This is confusing, even though the language allows it. I'll create a separate issue to implement this check as a part of Diktat.

That is very nice idea for inspection, Intellij analyzer does not trigger on that, but I am sure that it should be prohibited.

**Review by akuleshov7 — APPROVED** [review]
